Rubberduck.RibbonDispatcher/Sample Ribbons/RibbonLibrary/src/RibbonCallbacksAdapter.bas

```diff
@@ -0,0 +1,226 @@
+Attribute VB_Name = "RibbonCallbacksAdapter"
+Option Explicit
+
+Public Const ModuleName   As String = "RibbonCallbacksAdapter."
+Public Const DefaultImage As String = "MacroSecurity"
+
+Public Enum ControlSize
+    rdRegular = RdControlSize_rdRegular
+    rdLarge = RdControlSize_rdLarge
+End Enum
+
+Private Function ViewModelFor(ByVal Control As IRibbonControl _
+) As Rubberduck_RibbonDispatcher.RibbonViewModel
+    On Error GoTo EH
+    Dim WkBk As IRibbonWorkbook
+    Set WkBk = Control.Context.Parent
+    Set ViewModelFor = WkBk.ViewModel
+XT: Exit Function
+EH: If Err.Number <> 438 Then ReraiseError Err, ModuleName & "NewRibbonModel"
+    Resume XT
+End Function
+
+''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''
+'''                             Published Actions
+''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''
+Public Sub LoadImage(ByVal ImageId, ByRef Image)
+    On Error GoTo EH
+    Set Image = Nothing
+XT: Exit Sub
+EH: If Err.Number <> 91 Then DisplayError Err, ModuleName & "LoadImage", ImageId
+    Resume XT
+End Sub
+
+''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''
+'''                         Common to Many Controls
+''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''
+Public Sub GetEnabled(ByVal Control As IRibbonControl, ByRef Enabled)
+    On Error GoTo EH
+    Enabled = ViewModelFor(Control).GetEnabled(Control)
+XT: Exit Sub
+EH: If Err.Number <> 91 Then DisplayError Err, ModuleName & "GetEnabled", Control.Id
+    Resume XT
+End Sub
+Public Sub GetVisible(ByVal Control As IRibbonControl, ByRef Visible)
+    On Error GoTo EH
+    Visible = ViewModelFor(Control).GetVisible(Control)
+XT: Exit Sub
+EH: If Err.Number <> 91 Then DisplayError Err, ModuleName & "GetVisible", Control.Id
```

Comment:
So wait a minute, we deliberately access a potentially null reference, just so we have an error to handle?

How about this:

    Dim vm As Rubberduck_RibbonDispatcher.RibbonViewModel
    Set vm = ViewModelFor(Control)
    If Not vm Is Nothing Then Description = vm.GetDescription(Control)

No need for error handling here. Also seeing the API from a VBA standpoint, makes me wonder. `vm.Description` would look much cleaner - actually, this would be cleaner:

    If Not vm Is Nothing Then outResult = vm.Description

We already supply the `Control` to get the view model: it feels wrong to have to supply it again to get the description for it. Why can't the view model instance readily provide that string?

Reply:
Have you understood the problem being solved here, before assuming that a simpler writing works? Perhaps I originally wrote the code as you suggest, and that was an inadequate implementation? Perhaps the subsequent adoption of Qualified references in the Ribbon XML made the complex implementation redundant, or perhaps it didn't? Perhaps the test has simply not been done yet? Why do you make these simplistic demands before actually reading and understanding, instead of merely scanning, the code?